src/main/scala/is/hail/rvd/OrderedRVD.scala

```diff
@@ -37,40 +37,86 @@ class OrderedRVD(
+
   def boundary: OrderedRVD = OrderedRVD(typ, partitioner, crddBoundary)
 
   def rowType: TStruct = typ.rowType
 
   def updateType(newTyp: OrderedRVDType): OrderedRVD =
     OrderedRVD(newTyp, partitioner, crdd)
 
-  def mapPreservesPartitioning(newTyp: OrderedRVDType)(f: (RegionValue) => RegionValue): OrderedRVD =
+  def mapPreservesPartitioning(
+    newTyp: OrderedRVDType
+  )(f: (RegionValue) => RegionValue
+  ): OrderedRVD = {
+    require(newTyp.kType isPrefixOf typ.kType)
     OrderedRVD(newTyp,
-      partitioner,
+      partitioner.coarsen(newTyp.key.length),
       crdd.map(f))
+  }
 
-  def mapPartitionsWithIndexPreservesPartitioning(newTyp: OrderedRVDType)(f: (Int, Iterator[RegionValue]) => Iterator[RegionValue]): OrderedRVD =
-    OrderedRVD(newTyp,
-      partitioner,
+  def mapPartitionsWithIndexPreservesPartitioning(
+    newTyp: OrderedRVDType
+  )(f: (Int, Iterator[RegionValue]) => Iterator[RegionValue]
+  ): OrderedRVD = {
+    require(newTyp.kType isPrefixOf typ.kType)
+    OrderedRVD(
+      newTyp,
+      partitioner.coarsen(newTyp.key.length),
       crdd.mapPartitionsWithIndex(f))
+  }
 
   def mapPartitionsWithIndexPreservesPartitioning(
     newTyp: OrderedRVDType,
     f: (Int, RVDContext, Iterator[RegionValue]) => Iterator[RegionValue]
-  ): OrderedRVD = OrderedRVD(
-    newTyp,
-    partitioner,
-    crdd.cmapPartitionsWithIndex(f))
+  ): OrderedRVD = {
+    require(newTyp.kType isPrefixOf typ.kType)
+    OrderedRVD(
+      newTyp,
+      partitioner.coarsen(newTyp.key.length),
+      crdd.cmapPartitionsWithIndex(f))
+  }
 
-  def mapPartitionsPreservesPartitioning(newTyp: OrderedRVDType)(f: (Iterator[RegionValue]) => Iterator[RegionValue]): OrderedRVD =
-    OrderedRVD(newTyp,
-      partitioner,
+  def mapPartitionsPreservesPartitioning(
+    newTyp: OrderedRVDType
+  )(f: (Iterator[RegionValue]) => Iterator[RegionValue]
+  ): OrderedRVD = {
+    require(newTyp.kType isPrefixOf typ.kType)
+    OrderedRVD(
+      newTyp,
+      partitioner.coarsen(newTyp.key.length),
       crdd.mapPartitions(f))
+  }
 
   def mapPartitionsPreservesPartitioning(
     newTyp: OrderedRVDType,
     f: (RVDContext, Iterator[RegionValue]) => Iterator[RegionValue]
-  ): OrderedRVD = OrderedRVD(newTyp, partitioner, crdd.cmapPartitions(f))
+  ): OrderedRVD = {
+    require(newTyp.kType isPrefixOf typ.kType)
+    OrderedRVD(
+      newTyp,
+      partitioner.coarsen(newTyp.key.length),
+      crdd.cmapPartitions(f))
+  }
+
+  def extendKeyPreservesPartitioning(newKey: Array[String]): OrderedRVD = {
+    require(newKey startsWith typ.key)
+    require(newKey.forall(typ.rowType.fieldNames.contains))
+    val orvdType = new OrderedRVDType(typ.partitionKey, newKey, typ.rowType)
+    if (OrderedRVDPartitioner.isValid(orvdType.kType, partitioner.rangeBounds))
+      copy(typ = orvdType)
+    else
+      constrainToOrderedPartitioner(orvdType, partitioner.extendKey(orvdType.kType))
+  }
```

Comment:
Under what conditions will the new partitioner be invalid? In particular, it seems that the allowedOverlap will be the same or greater and the ordering on the new key type agrees with the old ordering.

Is the issue related to having interval end points longer than the currently known key where, under the new key type ordering, the partitions are revealed to not be ordered. For example, if we somehow had two partitions with these range bounds:

```
[ [1, 1, 1], [1, 2, 3] ]
[ [1, 2, 2], [1, 3, 0] ]
```

I think this is valid for key length 2, but it is not valid for key length 3. I don't understand how this situation could arise though. Why/how do I construct a partitioner with interval endpoints that are invalid under a longer type?

Reply:
You're on the right track. The extreme example is if `this` has `typ.key == List()`. Then the current partitioner has range bounds `[Row(), Row()]` for every partition. If we then add key fields, this partitioner will be invalid if there is more than one partition.

More generally, we're now allowing partitioners like `[ [1, 1], [1, 2] ], [ [1, 2], [1, 2] ], [ [1, 2], [1, 3] ]`. If we add a new key field, then keys starting with `[1, 2]` are spread across three partitions, and we can't tell if the third key field is ordered within those starting with `[1, 2]` without communication between partitions.